src/Calypso-SystemPlugins-Reflectivity-Browser-Tests/ClyInstallMetaLinkPresenterTest.class.st

```diff
@@ -0,0 +1,142 @@
+Class {
+	#name : #ClyInstallMetaLinkPresenterTest,
+	#superclass : #TestCase,
+	#instVars : [
+		'node',
+		'metalink',
+		'metalink2',
+		'breakpoint',
+		'executionCounter',
+		'watchpoint'
+	],
+	#category : #'Calypso-SystemPlugins-Reflectivity-Browser-Tests-Metalinks'
+}
+
+{ #category : #'helper methods' }
+ClyInstallMetaLinkPresenterTest >> dummyMethod [
+	self
+]
+
+{ #category : #'helper methods' }
+ClyInstallMetaLinkPresenterTest >> nodeInRealMethod [
+	^(self class >> #dummyMethod) ast statements first
+]
+
+{ #category : #running }
+ClyInstallMetaLinkPresenterTest >> presenterForMetalinkInstallation [
+	^ClyMetaLinkInstallationPresenter onNode: node forInstallation:  true
+]
+
+{ #category : #running }
+ClyInstallMetaLinkPresenterTest >> presenterForMetalinkUninstallation [
+	^ ClyMetaLinkInstallationPresenter onNode: node forInstallation: false
+]
+
+{ #category : #running }
+ClyInstallMetaLinkPresenterTest >> setUp [
+	super setUp.
+	MetaLink uninstallAll.
+	node := RBTemporaryNode named: 'test'.
+	metalink := MetaLink new.
+	metalink2 := MetaLink new.
+	breakpoint := MetaLink new metaObject: Break; yourself.
+	executionCounter := MetaLink new metaObject: ExecutionCounter new; yourself.
+	watchpoint := MetaLink new metaObject: Watchpoint new; yourself.
+	node propertyAt: #links put: {metalink. breakpoint. executionCounter. watchpoint} asOrderedCollection 
+	
+]
+
+{ #category : #running }
+ClyInstallMetaLinkPresenterTest >> tearDown [
+	metalink uninstall.
+	metalink2 uninstall.
+	super tearDown
+]
+
+{ #category : #tests }
+ClyInstallMetaLinkPresenterTest >> testInstallSelectedMetalink [
+	|presenter list|
+	presenter := ClyMetaLinkInstallationPresenter onNode: self nodeInRealMethod forInstallation:  true.
+	list := presenter instVarNamed: #list.
```

Comment:
Can we not use reflection to access variable? 
Otherwise variable references will never show this code.

Reply:
I do not understand.
Which variable do you mean?

Reply:
I mean instVarNamed: #list.

Reply:
It can be simple:
```Smalltalk
presenter list clickItem: 1
```

Reply:
which made me think that #list is not really meaningful name for variable

Reply:
Ah, yes, I understood "can we not...?" by "why not using", so it was not clear to me.
This #list is a dumb spec list presenter and is the only widget with a toolbar in the owner presenter.
I did not want to put an accessor just for the test, as it is not meant to be accessed.

Reply:
The problem is that such approach breaks tools:
- code navigation. User will browse this class and analyze #list variable. User will not find these tests using var references.
- refactoring. User will rename the #list variable. Tests will become broken because only variable is renamed, not a #list symbol

And most important it directly breaks incapsulation.   
If you need access to the variable from the outer of class you need an accessor. There is no other way.

Reply:
I'm not convinced encapsulation is that important when writing such UI tests, because it exposes my state only for tests and now to users it's part of the API while in reality it should not (and since reflection is there...).

But you are right it breaks all tools.

So I added an accessor, but then as you suggested I also changed the name (to #metalinkList) because now it needs to really mean something from a user point of view.
I realize now it should maybe be named metalinkListPresenter because it is not a list and it should be as an extension in the test package.

Reply:
Renamed to metalinkListPresenter